p2p/ipld/sample.go

```diff
@@ -0,0 +1,99 @@
+package ipld
+
+import (
+	"github.com/ipfs/go-cid"
+	"github.com/lazyledger/nmt/namespace"
+
+	"github.com/lazyledger/lazyledger-core/libs/rand"
+	"github.com/lazyledger/lazyledger-core/p2p/ipld/plugin/nodes"
+	"github.com/lazyledger/lazyledger-core/types"
+)
+
+// Sample is a point in 2D space over square.
+type Sample struct {
+	Row, Col uint32
+}
+
+// SampleSquare randomly picks *num* unique points from arbitrary *width* square
+// and returns them as samples.
+func SampleSquare(squareWidth uint32, num int) []*Sample {
+	ss := newSquareSampler(squareWidth, num)
+	ss.sample(num)
+	return ss.sampled()
+}
```

Comment:
I could have put all the sampling code into the one `SampleSquare` function without the introduction of `squareSampler`, however, I decided to go with the structure as we may decide to improve the random sampling algorithm for better sample picking distribution, e.g. by using [Poison-Disk Sampling](https://www.jasondavies.com/poisson-disc/), and having such structure will simplify things slightly.

Reply:
BTW, @adlerjohn, it might a good idea to include/describe random sampling algorithm into specs, as there are options that can non obviously affect sampling quality.

Reply:
Ref https://github.com/lazyledger/lazyledger-specs/issues/110